### PR TITLE
Add a plugin for individual-tree segmentation

### PR DESCRIPTION
**Pull request by truebelief**
Add a plugin for individual-tree segmentation (treeiso) with command-line support

TreeIso can be utilized to separate individual trees from terrestrial laser scanning point clouds, assigning tree IDs as supplementary scalar fields.

Detail of the methods can be found:
Xi, Z.; Hopkinson, C. 3D Graph-Based Individual-Tree Isolation (Treeiso) from Terrestrial Laser Scanning Point Clouds. Remote Sens. 2022, 14, 6116. https://doi.org/10.3390/rs14236116

The plugin depends on the following open-source libraries: [Boost (v1.67)](https://www.boost.org/users/history/version_1_67_0.html), [Eigen](https://eigen.tuxfamily.org/index.php?title=Main_Page), [cut-pursuit](https://github.com/loicland/cut-pursuit), [knn-cpp](https://github.com/Rookfighter/knn-cpp). 
All of these libraries are header-only. To reduce storage requirements, the original Boost library has been trimmed down to a minimal size of approximately 30MB, specifically for this plugin.

Currently, the plugin compiles and runs successfully using default settings on my PC (Windows 10, Visual Studio 2019), without further testing. In the next stage, the code formatting will be refined, and annotations will be added to elucidate the purpose and functionality of each major function.

The results of the segmentation can be viewed at the following repositories: [artemis_treeiso](https://github.com/truebelief/artemis_treeiso) and [artemis_treeiso (qTreeIso plugin)](https://github.com/truebelief/artemis_treeiso/tree/main/CloudCompare/plugins/qTreeIso).

This plugin is not designed to offer a flawless segmentation solution. Instead, it aims to divide segmentation into multiple levels of resolution, enabling users to edit and refine segmentation in conjunction with other existing tools in CloudCompare.

### CODE REVIEW

**Comment by truebelief** [comment]
Here is what the plugin looks like:
snapshot1
snapshot2
snapshot3
snapshot4

Thank you for consideration

**Review by rjanvier — CHANGES_REQUESTED** [review]
Hey thank you very much for this contribution. Working more or less In the same field currently it is a nice coincidence 👍 

-Hey thank you very much for this contribution. Working more or less In the same field currently it is a nice coincidence 👍 

- Maybe it's better if it stays a submodule, this way maintaining your package could be easier on your side. I'm not strongly opinionated on this but it is definitly something to discuss with @dgirardeau
- I would suggest not to embed dependencies but to use regular `find_package` mechanism for both boost and eigen. For example, currently your boost subset miss some headers and it does not compile as is. With find_pacakge for boost you can provides the minimal version and also the components you use (al least graph and geometry I think). I compiled your plugin this way on my side
- I would suggest not to embed dependencies but to use regular `find_package` mechanism for both boost and eigen. For example, currently your boost subset miss some headers and it does not compile as is. With find_pacakge for boost you can provides the minimal version and also the components you use (al least graph and geometry I think). I compiled your plugin this way on my side
- Maybe, for the KDTree implementation: we now embed a version of nanoflann in cccorelib, maybe you can reuse this and ditch knn-cpp. NanoFlann integrates well with Eigen and I think it could be faster and it is very well maintained.

**Comment by dgirardeau** [comment]
Wow, this plugin looks very interesting. Thanks for the contribution.

I agree to all of Romain's comments.

Regarding the first one: yes, it's clearly simpler/cleaner to keep your plugin where it is and integrate it as a submodule.

And for Eigen and Boost, it should definitely not be part of this repo, just let people specify where it's stored on their side (as other plugins are already doing, you can find examples for both Eigen and Boost in other plugins/parts of CC).

**Comment by dgirardeau** [comment]
@truebelief Will you be able to apply some changes?

**Comment by truebelief** [comment]
Yes, Daniel. Thank you and Romain for all these comments. They are all
operable. Sorry for the wait. I am traveling this month and I cannot
remotely debug the program with my laptop. I will find a time to update
this program by this weekend.


@truebelief <https://github.com/truebelief> Will you be able to apply some

**Comment by dgirardeau** [comment]
No hurry, I was just wondering if you would be able to update this PR at some point, or if we would have needed to do this ourselves! Take your time.

File: plugins/core/Standard/qTreeIso/include/API.h
[the remark this answers is not in the document]
Comment (truebelief) [reply plugins/core/Standard/qTreeIso/include/API.h:1]: I have added the MIT Licence for the cut-pursuit plugin to the root of the plugin, as you suggested. Thank you for pointing out the appropriate location for this file.

File: plugins/core/Standard/qTreeIso/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+option( PLUGIN_STANDARD_QTREEISO "Install qTreeIso plugin" OFF )
```
[the remark this answers is not in the document]
Comment (truebelief) [reply plugins/core/Standard/qTreeIso/CMakeLists.txt:1]: This was added. Thank you for the reminder.

**Comment by truebelief** [comment]
I just updated the files, and your suggestions helped make the program more structured. Thanks for the awesome advice.

Initially, I wanted to use nanoflann for the k-nearest neighbors search, but it kept giving me incorrect results when there were only a few points and when I tried reusing the nanoflann structure for searching. I'm not sure what went wrong, but thankfully, knn-cpp doesn't have the same issue.

I've been having a great time learning from this open-source project, and I'm looking forward to converting some other MATLAB codes, like 3D tree reconstruction algorithms, into CloudCompare plugins in the future.

> Hey thank you very much for this contribution. Working more or less In the same field currently it is a nice coincidence 👍
> 
> * Maybe it's better if it stays a submodule, this way maintaining your package could be easier on your side. I'm not strongly opinionated on this but it is definitly something to discuss with @dgirardeau
> * I would suggest not to embed dependencies but to use regular `find_package` mechanism for both boost and eigen. For example, currently your boost subset miss some headers and it does not compile as is. With find_pacakge for boost you can provides the minimal version and also the components you use (al least graph and geometry I think). I compiled your plugin this way on my side
> * I would suggest not to embed dependencies but to use regular `find_package` mechanism for both boost and eigen. For example, currently your boost subset miss some headers and it does not compile as is. With find_pacakge for boost you can provides the minimal version and also the components you use (al least graph and geometry I think). I compiled your plugin this way on my side
> * Maybe, for the KDTree implementation: we now embed a version of nanoflann in cccorelib, maybe you can reuse this and ditch knn-cpp. NanoFlann integrates well with Eigen and I think it could be faster and it is very well maintained.

**Comment by rjanvier** [comment]
Thank you so much, I think I'll have time to try and review the new version this WE.

- In the meantime could you please isolate the plugin and make it a dedicated repository. Please follow the example here https://github.com/hvs-ait/mplane-plugin/tree/342ef9f5c6c0068e313868496ce736ba69158a18. This way we could add it as a submodule to CC.
- If you have a code snippet to reproduce the issue with nanoflann I think I could have a look too.

**Comment by truebelief** [comment]
> Thank you so much, I think I'll have time to try and review the new version this WE.
> 
> * In the meantime could you please isolate the plugin and make it a dedicated repository. Please follow the example here https://github.com/hvs-ait/mplane-plugin/tree/342ef9f5c6c0068e313868496ce736ba69158a18. This way we could add it as a submodule to CC.
> * If you have a code snippet to reproduce the issue with nanoflann I think I could have a look too.

Thank you for all your comments and guidance. I've created a repository at https://github.com/truebelief/cc-treeiso-plugin, as you suggested.

I would appreciate any help you can provide with the nanoflann issue I'm encountering. There may be an issue with my nanoflann settings.

I've compared the nanoflann results with the regular KNN search results, exporting the returned indices and distances to a text file. The regular KNN search produced the same results as Matlab's knnsearch, with the following code:

```
clusterCentroids=[[12.1,21.1,31.1];[12.1,21.1,30.1];[12.7667,21.4333,30.7667];[11.7667,20.4333,31.4333];[10.1,20.1,30.1];[10.1,20.1,30.1]];
[idx,D]=knnsearch(clusterCentroids,clusterCentroids,"K",4);
```

The indices returned by nanoflann were correct, but the distances were not. I recall encountering another issue with nanoflann during my debugging process. I'll work on simplifying the code related to that issue once we address the current problem.

```


#include <numeric>
#include <vector>
#include <string>

#include <fstream>
#include <iostream>


#include "nanoflann.hpp"
#include "KDTreeVectorOfVectorsAdaptor.h"

using namespace std;



//90 % of this function was generated by the ChatGPT!
template <typename T1>
bool nanoflann_knnsearch(std::vector<std::vector<T1>>& data, std::vector<std::vector<T1>>& query, unsigned K, std::vector<std::vector<size_t>>& nn_idx, std::vector<std::vector<float>>& nn_distance) {
	nn_idx.clear();
	nn_distance.clear();
	typedef KDTreeVectorOfVectorsAdaptor <std::vector<std::vector<T1>>, T1> KDTreeVV;
	KDTreeVV kdtree_index(3, data);
	std::vector<size_t> indices(K);
	std::vector<float> distances(K);

	for (auto q : query) {
		kdtree_index.index->knnSearch(&q[0], K, &indices[0], &distances[0]);
		nn_idx.push_back(indices);
		nn_distance.push_back(distances);
	}
	return true;
}


// Compute the Euclidean distance between two points
float euclidean_distance(const std::vector<float>& a, const std::vector<float>& b) {
	float sum = 0.0;
	for (size_t i = 0; i < a.size(); ++i) {
		float diff = a[i] - b[i];
		sum += diff * diff;
	}
	return sqrt(sum);
}

// Find the k nearest neighbors of a query point in a dataset
void k_nearest_neighbors_per_query(const std::vector<std::vector<float>>& dataset,
	const std::vector<float>& query_point, size_t k, std::vector<size_t>& neighbor, std::vector<float>& distance) {
	// Create a vector of pairs to store the distances and indices of each point in the dataset
	std::vector<std::pair<float, size_t>> distances_and_indices(dataset.size());
	for (size_t i = 0; i < dataset.size(); ++i) {
		float distance = euclidean_distance(dataset[i], query_point);
		distances_and_indices[i] = std::make_pair(distance, i);
	}

	// Sort the distances and indices in ascending order of distance
	sort(distances_and_indices.begin(), distances_and_indices.end());

	// Return the indices of the k nearest neighbors
	size_t minK = (distances_and_indices.size() > k ? k : distances_and_indices.size());
	std::vector<size_t> nbr(minK);
	std::vector<float> dist(minK);

	for (size_t i = 0; i < minK; ++i) {
		nbr[i] = distances_and_indices[i].second;
		dist[i] = distances_and_indices[i].first;
	}
	neighbor = nbr;
	distance = dist;
	return;
}


// Find the k nearest neighbors of a query point in a dataset
void k_nearest_neighbors(const std::vector<std::vector<float>>& dataset,
	const std::vector <std::vector<float>>& query_points, size_t k, std::vector <std::vector<size_t>>& neighbors, std::vector <std::vector<float>>& distances) {
	// Create a vector of pairs to store the distances and indices of each point in the dataset
	for (size_t i = 0; i < query_points.size(); ++i) {
		std::vector<size_t>neighbor;
		std::vector<float>distance;
		k_nearest_neighbors_per_query(dataset, query_points[i], k, neighbor, distance);
		neighbors.push_back(neighbor);
		distances.push_back(distance);
	}

	return;
}

template <typename T>
void write_log(std::vector< std::vector<T>>arr, std::string fname) {
	std::ofstream outfile(fname);
	for (const auto& row : arr) {
		for (const auto& elem : row) {
			outfile << elem << " ";
		}
		outfile << std::endl;
	}
	outfile.close();
}


int main()
{
	const unsigned int K = 3;
	std::vector<std::vector<float>> clusterCentroids = {{12.1,21.1,31.1},{12.1,21.1,30.1},{12.7667,21.4333,30.7667},{11.7667,20.4333,31.4333},{10.1,20.1,30.1},{10.1,20.1,30.1}};//Create a test 2D array data
	write_log(clusterCentroids, "clusterCentroids.txt");


	std::vector<std::vector<size_t>> minIdxsC;
	std::vector<std::vector<float>> minIdxsD;
	nanoflann_knnsearch(clusterCentroids, clusterCentroids, K, minIdxsC, minIdxsD);//Run nanoflann
	write_log(minIdxsC, "nanoflann_idx.txt");
	write_log(minIdxsD, "nanoflann_D.txt");

	std::vector<std::vector<size_t>> minIdxsC1;
	std::vector<std::vector<float>> minIdxsD1;
	k_nearest_neighbors(clusterCentroids, clusterCentroids, K, minIdxsC1, minIdxsD1);//Run regular KNN
	write_log(minIdxsC1, "knn_idx.txt");
	write_log(minIdxsD1, "knn_D.txt");

}

```

**Comment by truebelief** [comment]
Now that I've set up a plugin repository at https://github.com/truebelief/cc-treeiso-plugin, should we close this pull request and move future discussions there?

I recently wrapped up a lab session introducing students and guests to basic CloudCompare functionalities. I explored fundamental tasks such as measuring tree stem diameter at specific heights, which was achieved using the "extract sections" tool. But the main focus was on fitting a circle/cylinder into the height-specific slice. I found the "Ransac shape detection" tool to be less effective for sparse points, requiring a lot of trial and error without giving the desired results. As a workaround, I used the "fit sphere" tool, which sometimes caused overestimations.

I've been mulling over the idea of extending the existing geometry fitting tools to include a circle fitting algorithm. I believe it's something that many forestry professionals could benefit from and it's been notably absent for a while. I'm eager to hear from you or Daniel on how to best contribute such code to the repository. Would it be better to create a separate repository, similar to what I did for the plugin, or should I simply open a new pull request?

**Comment by dgirardeau** [comment]
Yes, we can close this PR but you'll still need to make a new PR to add the plugin as a submodule. I can also take care of that if you prefer.

And for a circle fitting tool, I believe it's universal enough to add this to the main project. No need to make a plugin for that.

**Comment by dgirardeau** [comment]
Hum, not sure why you have 'merged' and old version of CC's master branch in yours, but now it's in a dire state, and we won't be able to merge anything from this branch... I'll try to save what I can and create a new PR.

**Comment by dgirardeau** [comment]
So I've made a major cleanup here: https://github.com/truebelief/cc-treeiso-plugin/pull/2

**Comment by truebelief** [comment]
> Hum, not sure why you have 'merged' and old version of CC's master branch in yours, but now it's in a dire state, and we won't be able to merge anything from this branch... I'll try to save what I can and create a new PR.

Sorry, I was working on local files separate from the GitHub folders and did not pay attention to the versions of CC. Next time I will work on the synchronized GitHub files directly.

**Comment by dgirardeau** [comment]
Here is the new PR to add the plugin in a clean way: https://github.com/CloudCompare/CloudCompare/pull/1785

**Pull request closed without merging**